File: parser.py
import mistune
import re


def parse_spaceup(input_str):
    """Parse Spaceup markup into HTML."""
    lines = [line.rstrip() for line in input_str.splitlines()]  # Clean trailing whitespace
    indent_stack = [0]  # Start with root level 0
    output = []
    pos = 0
    previous_non_whitespace_indent = 0
    
    def compute_indent(line):
        stripped = line.lstrip()
        if not stripped or stripped.startswith('//'):  # Ignore blank or pure comment lines for structure
            return None
        return len(line) - len(stripped)
    
    def peek_next_indent(current_pos):
        for i in range(current_pos + 1, len(lines)):
            indent = compute_indent(lines[i])
            if indent is not None:
                return indent
        return -1  # EOF, use -1 to indicate
    
    def render_inline_markdown(text: str) -> str:
        html = mistune.html(text).strip()
        if html.startswith('<p>') and html.endswith('</p>'):
            html = html[3:-4]
        return html

    def emit_paragraph(text_lines):
        if not text_lines:
            return
        output.append('<p>')
        for item in text_lines:
            if isinstance(item, tuple):
                text, comment = item
                rendered = render_inline_markdown(text)
                if comment:
                    output.append(f'{rendered}  <!-- {comment} --><br>')
                else:
                    output.append(f'{rendered}<br>')
            else:
                output.append(f'{render_inline_markdown(str(item))}<br>')
        output.append('</p>')

    def emit_list(items):
        if not items:
            return
        lis = ''
        for text in items:
            if text.startswith('[ ] '):
                lis += '<li><input type="checkbox" disabled> ' + render_inline_markdown(text[4:]) + '</li>\n'
            elif text.startswith('[x] '):
                lis += '<li><input type="checkbox" checked disabled> ' + render_inline_markdown(text[4:]) + '</li>\n'
            else:
                lis += '<li>' + render_inline_markdown(text) + '</li>\n'
        output.append('<ul>')
        output.append(lis)
        output.append('</ul>')

    def emit_ordered_list(items):
        if not items:
            return
        lis = '\n'.join(f'<li>{render_inline_markdown(text)}</li>' for text in items)
        output.append('<ol>')
        output.append(lis)
        output.append('</ol>')

    def emit_code_block(language, code_lines):
        if not code_lines:
            return
        code_content = '\n'.join(code_lines)
        output.append(f'<pre><code class="language-{language or ""}">{code_content}</code></pre>')

    def emit_blockquote(lines):
        if not lines:
            return
        combined = '\n'.join(line for line in lines)
        html = mistune.html(combined)
        output.append(html)

    def emit_table(rows):
        if not rows or len(rows) < 3:  # Min header, separator, one row
            return
        output.append('<table>')
        # Header
        output.append('<thead><tr>')
        for cell in rows[0].split('|')[1:-1]:
            output.append('<th>' + cell.strip() + '</th>')
        output.append('</tr></thead>')
        # Body
        output.append('<tbody>')
        for row in rows[2:]:
            output.append('<tr>')
            for cell in row.split('|')[1:-1]:
                output.append('<td>' + cell.strip() + '</td>')
            output.append('</tr>')
        output.append('</tbody>')
        output.append('</table>')

    def split_content_and_inline_comment(text):
        stripped = text.lstrip()
        search_start = 0
        while True:
            idx = stripped.find('//', search_start)
            if idx == -1:
                break
            if idx == 0:
                # Comment-only lines are handled elsewhere; skip here
                search_start = idx + 2
                continue
            if stripped[idx - 1].isspace():
                return stripped[:idx].rstrip(), stripped[idx + 2 :].strip()
            search_start = idx + 2
        return stripped.strip(), None

    def extract_comment_only(text):
        stripped = text.lstrip()
        if stripped.startswith('//'):
            return stripped[2:].strip()
        return None

    def parse_element(current_indent):
        nonlocal pos
        nonlocal previous_non_whitespace_indent
        while pos < len(lines):
            # Skip blanks, but emit comment-only lines as HTML comments
            while pos < len(lines):
                stripped = lines[pos].lstrip()
                if not stripped:
                    pos += 1
                    continue
                if stripped.startswith('//'):
                    comment_text = stripped[2:].strip()
                    if comment_text:
                        output.append(f'<!-- {comment_text} -->')
                    pos += 1
                    continue
                break
            
            if pos >= len(lines):
                return
            
            line = lines[pos]
            indent = compute_indent(line)
            if indent < current_indent:
                return  # Dedent
            
            # Extract content and inline comment (if present)
            content, inline_comment = split_content_and_inline_comment(line)
            if not content:
                pos += 1
                continue  # Skip if no content after stripping
            
            # Detect fenced code blocks early, before heading/paragraph logic
            if content.startswith('```'):
                # Extract language if present
                language = content[3:].strip() or None
                code_lines = []
                pos += 1
                while pos < len(lines):
                    next_line = lines[pos]
                    if next_line.strip().startswith('```'):
                        pos += 1
                        break
                    code_lines.append(next_line[indent:])  # Trim exactly to the opening indent level
                    pos += 1
                # Dedent the code block
                if code_lines:
                    min_indent = min(len(l) - len(l.lstrip()) for l in code_lines if l.strip())
                    code_lines = [l[min_indent:] for l in code_lines]
                emit_code_block(language, code_lines)
                previous_non_whitespace_indent = indent
                continue

            next_indent = peek_next_indent(pos)
            has_blank_before_next = False
            is_followed_by_code_block = False
            for i in range(pos + 1, len(lines)):
                stripped = lines[i].lstrip()
                if not stripped or stripped.startswith('//'):
                    if stripped:  # Non-empty blank or comment counts as separator
                        has_blank_before_next = True
                    continue
                if stripped.startswith('```'):
                    is_followed_by_code_block = True
                break
            ambiguous_decrease = (
                previous_non_whitespace_indent > indent and next_indent == indent and not has_blank_before_next
            )
            is_heading = next_indent > indent or (next_indent == indent and has_blank_before_next) or ambiguous_decrease or is_followed_by_code_block or (next_indent == -1 and pos < len(lines) - 1)

            heading_level = len(indent_stack)
            if is_heading:
                rendered_heading = render_inline_markdown(content)
                output.append(f'<h{heading_level}>{rendered_heading}</h{heading_level}>')
                previous_non_whitespace_indent = indent
                indent_stack.append(indent)
                pos += 1
                # Handle ambiguous decrease: force subsequent same-indented lines as children
                if ambiguous_decrease:
                    forced_para_lines = []
                    while pos < len(lines):
                        nindent = compute_indent(lines[pos])
                        if nindent is None or nindent != indent:
                            break
                        ncontent, ncomment = split_content_and_inline_comment(lines[pos])
                        if ncontent:
                            forced_para_lines.append((ncontent, ncomment))
                            previous_non_whitespace_indent = indent
                        pos += 1
                    emit_paragraph(forced_para_lines)
                parse_element(indent + 1)  # Recurse for content with greater indent
                indent_stack.pop()
            else:
                # Paragraph: collect consecutive lines at same indent
                # But first, detect unordered list starting with "- "
                if content.startswith('- '):
                    items = [content[2:].strip()]
                    pos += 1
                    while pos < len(lines):
                        next_line_indent = compute_indent(lines[pos])
                        if next_line_indent != indent:
                            break
                        next_content, _next_comment = split_content_and_inline_comment(lines[pos])
                        if not next_content.startswith('- '):
                            break
                        items.append(next_content[2:].strip())
                        pos += 1
                    emit_list(items)
                    previous_non_whitespace_indent = indent
                    continue

                elif re.match(r'^\d+\.\s', content):
                    items = []
                    match = re.match(r'^\d+\.\s(.*)', content)
                    if match:
                        items.append(match.group(1).strip())
                    pos += 1
                    while pos < len(lines):
                        next_line_indent = compute_indent(lines[pos])
                        if next_line_indent != indent:
                            break
                        next_content, _next_comment = split_content_and_inline_comment(lines[pos])
                        match = re.match(r'^\d+\.\s(.*)', next_content)
                        if not match:
                            break
                        items.append(match.group(1).strip())
                        pos += 1
                    emit_ordered_list(items)
                    previous_non_whitespace_indent = indent
                    continue

                elif content.startswith('> '):
                    quote_lines = [content]
                    pos += 1
                    while pos < len(lines):
                        next_line_indent = compute_indent(lines[pos])
                        if next_line_indent != indent:
                            break
                        next_content, _ = split_content_and_inline_comment(lines[pos])
                        if not next_content.startswith('> '):
                            break
                        quote_lines.append(next_content)
                        pos += 1
                    emit_blockquote(quote_lines)
                    previous_non_whitespace_indent = indent
                    continue

                elif content.startswith('|') and '|' in content[1:]:
                    table_lines = [content]
                    pos += 1
                    while pos < len(lines):
                        next_line_indent = compute_indent(lines[pos])
                        if next_line_indent != indent:
                            break
                        next_content, _ = split_content_and_inline_comment(lines[pos])
                        if not next_content.startswith('|') or '|' not in next_content[1:]:
                            break
                        table_lines.append(next_content)
                        pos += 1
                    emit_table(table_lines)
                    previous_non_whitespace_indent = indent
                    continue

                para_lines = [(content, inline_comment)]
                pos += 1
                while pos < len(lines):
                    next_line_indent = compute_indent(lines[pos])
                    if next_line_indent != indent:
                        break
                    next_content, next_comment = split_content_and_inline_comment(lines[pos])
                    if not next_content:
                        pos += 1
                        continue
                    para_lines.append((next_content, next_comment))
                    previous_non_whitespace_indent = indent
                    pos += 1
                emit_paragraph(para_lines)
                previous_non_whitespace_indent = indent
    
    parse_element(0)
    return '\n'.join(output)

File: rs/src/parser.rs
use pulldown_cmark::{Options, Parser as MdParser, html};
use regex::Regex;

pub fn parse_spaceup(input: &str) -> String {
	let lines: Vec<String> = input.lines().map(|l| l.trim_end().to_string()).collect();
	let mut output: Vec<String> = Vec::new();
	let mut pos: usize = 0;
	let mut indent_stack: Vec<usize> = vec![0];
	let mut previous_non_ws_indent: usize = 0;

	fn compute_indent(line: &str) -> Option<usize> {
		let stripped = line.trim_start();
		if stripped.is_empty() { return None; }
		if stripped.starts_with("//") { return None; }
		Some(line.len() - stripped.len())
	}

	fn peek_next_indent(lines: &[String], current_pos: usize) -> isize {
		for i in current_pos+1..lines.len() {
			if let Some(ind) = compute_indent(&lines[i]) { return ind as isize; }
		}
		-1
	}

	fn render_inline_markdown(text: &str) -> String {
		let mut options = Options::empty();
		options.insert(Options::ENABLE_TABLES);
		options.insert(Options::ENABLE_TASKLISTS);
		let parser = MdParser::new_ext(text, options);
		let mut html_output = String::new();
		html::push_html(&mut html_output, parser);
		let html_output = html_output.trim().to_string();
		if html_output.starts_with("<p>") && html_output.ends_with("</p>") {
			html_output[3..html_output.len()-4].to_string()
		} else {
			html_output
		}
	}

	fn emit_paragraph(output: &mut Vec<String>, text_lines: &[(String, Option<String>)]) {
		if text_lines.is_empty() { return; }
		output.push("<p>".to_string());
		for (text, comment) in text_lines.iter() {
			let rendered = render_inline_markdown(text);
			if let Some(c) = comment {
				output.push(format!("    {}  <!-- {} --><br>", rendered, c));
			} else {
				output.push(format!("    {}<br>", rendered));
			}
		}
		output.push("</p>".to_string());
	}

	fn emit_list(output: &mut Vec<String>, items: &[String]) {
		if items.is_empty() { return; }
		output.push("<ul>".to_string());
		for item in items {
			let mut text = item.as_str();
			let line = if let Some(rest) = text.strip_prefix("[ ] ") {
				format!("    <li><input type=\"checkbox\" disabled> {}</li>", render_inline_markdown(rest))
			} else if let Some(rest) = text.strip_prefix("[x] ") {
				format!("    <li><input type=\"checkbox\" checked disabled> {}</li>", render_inline_markdown(rest))
			} else {
				format!("    <li>{}</li>", render_inline_markdown(text))
			};
			output.push(line);
		}
		output.push("</ul>".to_string());
	}

	fn emit_ordered_list(output: &mut Vec<String>, items: &[String]) {
		if items.is_empty() { return; }
		output.push("<ol>".to_string());
		for item in items {
			output.push(format!("    <li>{}</li>", render_inline_markdown(item)));
		}
		output.push("</ol>".to_string());
	}

	fn emit_code_block(output: &mut Vec<String>, language: Option<&str>, code_lines: &[String]) {
		if code_lines.is_empty() { return; }
		let code_content = code_lines.join("\n");
		let lang = language.unwrap_or("");
		output.push(format!("<pre><code class=\"language-{}\">{}</code></pre>", lang, code_content));
	}

	fn emit_blockquote(output: &mut Vec<String>, lines: &[String]) {
		if lines.is_empty() { return; }
		let combined = lines.join("\n");
		let mut options = Options::empty();
		let parser = MdParser::new_ext(&combined, options);
		let mut html_output = String::new();
		html::push_html(&mut html_output, parser);
		output.push(html_output);
	}

	fn emit_table(output: &mut Vec<String>, rows: &[String]) {
		if rows.len() < 3 { return; }
		output.push("<table>".to_string());
		output.push("<thead><tr>".to_string());
		for cell in rows[0].split('|').skip(1).take_while(|c| !c.is_empty()) {
			output.push(format!("<th>{}</th>", cell.trim()));
		}
		output.push("</tr></thead>".to_string());
		output.push("<tbody>".to_string());
		for row in rows.iter().skip(2) {
			output.push("<tr>".to_string());
			for cell in row.split('|').skip(1).take_while(|c| !c.is_empty()) {
				output.push(format!("<td>{}</td>", cell.trim()));
			}
			output.push("</tr>".to_string());
		}
		output.push("</tbody>".to_string());
		output.push("</table>".to_string());
	}

	fn split_content_and_inline_comment(text: &str) -> (String, Option<String>) {
		let stripped = text.trim_start();
		let mut search_start = 0usize;
		let bytes = stripped.as_bytes();
		while let Some(idx) = stripped[search_start..].find("//") {
			let i = search_start + idx;
			if i == 0 { search_start = i + 2; continue; }
			if bytes[i-1].is_ascii_whitespace() {
				let content = stripped[..i].trim_end().to_string();
				let comment = stripped[i+2..].trim().to_string();
				return (content, if comment.is_empty() { None } else { Some(comment) });
			}
			search_start = i + 2;
		}
		(stripped.trim().to_string(), None)
	}

	fn extract_comment_only(text: &str) -> Option<String> {
		let stripped = text.trim_start();
		if stripped.starts_with("//") {
			Some(stripped[2..].trim().to_string())
		} else { None }
	}

	fn parse_element(lines: &[String], pos_ref: &mut usize, current_indent: usize, indent_stack: &mut Vec<usize>, previous_non_ws_indent: &mut usize, output: &mut Vec<String>) {
		while *pos_ref < lines.len() {
			// skip blanks and emit comment-only lines
			while *pos_ref < lines.len() {
				let stripped = lines[*pos_ref].trim_start();
				if stripped.is_empty() { *pos_ref += 1; continue; }
				if stripped.starts_with("//") {
					let comment_text = stripped[2..].trim();
					if !comment_text.is_empty() { output.push(format!("<!-- {} -->", comment_text)); }
					*pos_ref += 1;
					continue;
				}
				break;
			}
			if *pos_ref >= lines.len() { return; }
			let line = &lines[*pos_ref];
			let indent = match compute_indent(line) { Some(v) => v, None => { *pos_ref += 1; continue } };
			if indent < current_indent { return; }

			let (content, inline_comment) = split_content_and_inline_comment(line);
			if content.is_empty() { *pos_ref += 1; continue; }

			// fenced code block
			if content.starts_with("```") {
				let language = content[3..].trim();
				let language = if language.is_empty() { None } else { Some(language) };
				let mut code_lines: Vec<String> = Vec::new();
				*pos_ref += 1;
				while *pos_ref < lines.len() {
					let next_line = &lines[*pos_ref];
					if next_line.trim().starts_with("```") { *pos_ref += 1; break; }
					code_lines.push(next_line[indent..].to_string());
					*pos_ref += 1;
				}
				// dedent
				if !code_lines.is_empty() {
					let min_indent = code_lines.iter()
						.filter(|l| !l.trim().is_empty())
						.map(|l| l.len() - l.trim_start().len())
						.min().unwrap_or(0);
					for l in code_lines.iter_mut() { if l.len() >= min_indent { *l = l[min_indent..].to_string(); } }
				}
				emit_code_block(output, language, &code_lines);
				*previous_non_ws_indent = indent;
				continue;
			}

			let next_indent = peek_next_indent(lines, *pos_ref);
			let mut has_blank_before_next = false;
			let mut is_followed_by_code_block = false;
			for i in *pos_ref + 1..lines.len() {
				let stripped = lines[i].trim_start();
				if stripped.is_empty() || stripped.starts_with("//") {
					if !stripped.is_empty() { has_blank_before_next = true; }
					continue;
				}
				if stripped.starts_with("```") { is_followed_by_code_block = true; }
				break;
			}
			let ambiguous_decrease = *previous_non_ws_indent > indent && next_indent == indent as isize && !has_blank_before_next;
			let is_heading = next_indent > indent as isize
				|| (next_indent == indent as isize && has_blank_before_next)
				|| ambiguous_decrease
				|| is_followed_by_code_block
				|| (next_indent == -1 && *pos_ref < lines.len() - 1);

			let heading_level = indent_stack.len();
			if is_heading {
				let rendered_heading = render_inline_markdown(&content);
				output.push(format!("<h{}>{}</h{}>", heading_level, rendered_heading, heading_level));
				*previous_non_ws_indent = indent;
				indent_stack.push(indent);
				*pos_ref += 1;
				if ambiguous_decrease {
					let mut forced_para_lines: Vec<(String, Option<String>)> = Vec::new();
					while *pos_ref < lines.len() {
						let nindent = compute_indent(&lines[*pos_ref]);
						if nindent.is_none() || nindent.unwrap() != indent { break; }
						let (ncontent, ncomment) = split_content_and_inline_comment(&lines[*pos_ref]);
						if !ncontent.is_empty() {
							forced_para_lines.push((ncontent, ncomment));
							*previous_non_ws_indent = indent;
						}
						*pos_ref += 1;
					}
					emit_paragraph(output, &forced_para_lines);
				}
				parse_element(lines, pos_ref, indent + 1, indent_stack, previous_non_ws_indent, output);
				indent_stack.pop();
			} else {
				// paragraph block or other structures at same indent
				let stripped = content.trim_start();
				if stripped.starts_with("- ") {
					let mut items = vec![stripped[2..].trim().to_string()];
					*pos_ref += 1;
					while *pos_ref < lines.len() {
						let next_line_indent = compute_indent(&lines[*pos_ref]);
						if next_line_indent != Some(indent) { break; }
						let (next_content, _) = split_content_and_inline_comment(&lines[*pos_ref]);
						if !next_content.trim_start().starts_with("- ") { break; }
						items.push(next_content.trim_start()[2..].trim().to_string());
						*pos_ref += 1;
					}
					emit_list(output, &items);
					*previous_non_ws_indent = indent;
					continue;
				} else if Regex::new(r"^\d+\.\s").unwrap().is_match(stripped) {
					let mut items: Vec<String> = Vec::new();
					if let Some(cap) = Regex::new(r"^\d+\.\s(.*)").unwrap().captures(stripped) {
						items.push(cap.get(1).unwrap().as_str().trim().to_string());
					}
					*pos_ref += 1;
					while *pos_ref < lines.len() {
						let next_line_indent = compute_indent(&lines[*pos_ref]);
						if next_line_indent != Some(indent) { break; }
						let (next_content, _) = split_content_and_inline_comment(&lines[*pos_ref]);
						if !Regex::new(r"^\d+\.\s").unwrap().is_match(next_content.trim_start()) { break; }
						if let Some(cap) = Regex::new(r"^\d+\.\s(.*)").unwrap().captures(next_content.trim_start()) {
							items.push(cap.get(1).unwrap().as_str().trim().to_string());
						}
						*pos_ref += 1;
					}
					emit_ordered_list(output, &items);
					*previous_non_ws_indent = indent;
					continue;
				} else if stripped.starts_with("> ") {
					let mut quote_lines = vec![content.clone()];
					*pos_ref += 1;
					while *pos_ref < lines.len() {
						let next_line_indent = compute_indent(&lines[*pos_ref]);
						if next_line_indent != Some(indent) { break; }
						let (next_content, _) = split_content_and_inline_comment(&lines[*pos_ref]);
						if !next_content.starts_with("> ") { break; }
						quote_lines.push(next_content);
						*pos_ref += 1;
					}
					emit_blockquote(output, &quote_lines);
					*previous_non_ws_indent = indent;
					continue;
				} else if stripped.starts_with('|') && stripped[1..].contains('|') {
					let mut table_lines = vec![content.clone()];
					*pos_ref += 1;
					while *pos_ref < lines.len() {
						let next_line_indent = compute_indent(&lines[*pos_ref]);
						if next_line_indent != Some(indent) { break; }
						let (next_content, _) = split_content_and_inline_comment(&lines[*pos_ref]);
						if !(next_content.starts_with('|') && next_content[1..].contains('|')) { break; }
						table_lines.push(next_content);
						*pos_ref += 1;
					}
					emit_table(output, &table_lines);
					*previous_non_ws_indent = indent;
					continue;
				} else {
					let mut para_lines: Vec<(String, Option<String>)> = vec![(content, inline_comment)];
					*pos_ref += 1;
					while *pos_ref < lines.len() {
						let next_line_indent = compute_indent(&lines[*pos_ref]);
						if next_line_indent != Some(indent) { break; }
						let (next_content, next_comment) = split_content_and_inline_comment(&lines[*pos_ref]);
						if next_content.is_empty() { *pos_ref += 1; continue; }
						para_lines.push((next_content, next_comment));
						*previous_non_ws_indent = indent;
						*pos_ref += 1;
					}
					emit_paragraph(output, &para_lines);
					*previous_non_ws_indent = indent;
				}
			}
		}
	}

	parse_element(&lines, &mut pos, 0, &mut indent_stack, &mut previous_non_ws_indent, &mut output);
	output.join("\n")
}